[GStreamer][VideoCapture] fix getVideoSizeAndFormatFromCaps/getVideoResolutionFromCaps for video/x-raw mime-type caps without format

jameshilliard opened the pull request:
#### fdb0e95ecfc0a8c011ed589562dd4162551d0f5a
<pre>
[GStreamer][VideoCapture] fix getVideoSizeAndFormatFromCaps/getVideoResolutionFromCaps for video/x-raw mime-type caps without format
<a href="https://bugs.webkit.org/show_bug.cgi?id=242290">https://bugs.webkit.org/show_bug.cgi?id=242290</a>

Reviewed by Philippe Normand.

Fixes:
0:00:02.001539663  1169 0x5577d2b74810 DEBUG             video-info video-info.c:406:gst_video_info_from_caps: parsing caps video/x-raw
0:00:02.001842884  1169 0x5577d2b74810 ERROR             video-info video-info.c:546:gst_video_info_from_caps: no format given
0:00:02.002197896  1169 0x5577d2b74810 INFO     webkitvideocapturer GStreamerVideoCapturer.cpp:136:setSize: Setting size to 640x480

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::getVideoSizeAndFormatFromCaps):
(WebCore::getVideoResolutionFromCaps):

Canonical link: <a href="https://commits.webkit.org/252112@main">https://commits.webkit.org/252112@main</a>
</pre>

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -126,7 +126,7 @@ bool getVideoSizeAndFormatFromCaps(const GstCaps* caps, WebCore::IntSize& size,
         size.setWidth(width);
         size.setHeight(height);
-    } else {
+    } else if (!gst_structure_has_name(structure, "video/x-raw") || gst_structure_has_name(structure, "format")) {
```
In which case did you see a structure named `format`?

jameshilliard replied:
The failure comes from [format being missing from here](https://github.com/GStreamer/gstreamer/blob/1.20.2/subprojects/gst-plugins-base/gst-libs/gst/video/video-info.c#L410) so checking `gst_structure_has_name(structure, "format")` guards against `gst_video_info_from_caps` being called in cases where `gst_video_info_from_caps` will fail due to not having `format`.

philn replied:
`gst_structure_get_string` fetches a field of the struct, not its name.

jameshilliard replied:
Yeah, just noticed that, changed to `gst_structure_has_field`.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -138,6 +138,19 @@ bool getVideoSizeAndFormatFromCaps(const GstCaps* caps, WebCore::IntSize& size,
+        format = GST_VIDEO_FORMAT_UNKNOWN;
+        stride = 0;
+        int width = 0, height = 0;
+        gst_structure_get_int(structure, "width", &width);
+        gst_structure_get_int(structure, "height", &height);
```
Return values should be checked.

jameshilliard replied:
Hmm, these fallbacks are based off of the `areEncryptedCaps` paths which also don't have checked returns:
https://github.com/WebKit/WebKit/blob/16de8a90c10880777b7e106e3e11b313cca02951/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp#L120-L121

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -153,12 +166,12 @@ std::optional<FloatSize> getVideoResolutionFromCaps(const GstCaps* caps)
         gst_structure_get_int(structure, "height", &height);
         gst_structure_get_fraction(structure, "pixel-aspect-ratio", &pixelAspectRatioNumerator, &pixelAspectRatioDenominator);
-    } else {
+    } else if (!gst_structure_has_name(structure, "video/x-raw") || gst_structure_has_name(structure, "format")) {
```
Ditto. Did you mean to use `gst_structure_has_field()` here?

jameshilliard replied:
Oh, probably.

jameshilliard replied:
Changed to `gst_structure_has_field`.

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -168,6 +181,10 @@ std::optional<FloatSize> getVideoResolutionFromCaps(const GstCaps* caps)
         pixelAspectRatioNumerator = GST_VIDEO_INFO_PAR_N(&info);
         pixelAspectRatioDenominator = GST_VIDEO_INFO_PAR_D(&info);
+    } else {
+        gst_structure_get_int(structure, "width", &width);
+        gst_structure_get_int(structure, "height", &height);
+        gst_structure_get_fraction(structure, "pixel-aspect-ratio", &pixelAspectRatioNumerator, &pixelAspectRatioDenominator);
```
Return values should be checked.

jameshilliard replied:
They don't seem to be getting checked elsewhere, not sure if that's actually needed here?:
https://github.com/WebKit/WebKit/blob/16de8a90c10880777b7e106e3e11b313cca02951/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp#L158-L160

jameshilliard commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -112,21 +112,8 @@ bool getVideoSizeAndFormatFromCaps(const GstCaps* caps, WebCore::IntSize& size,
-        size.setHeight(height);
-    } else {
+    GstStructure* structure = gst_caps_get_structure(caps, 0);
+    if (!areEncryptedCaps(caps) && (!gst_structure_has_name(structure, "video/x-raw") || gst_structure_has_field(structure, "format"))) {
```
Refactored this a bit to combine the encrypted caps codepaths with the no format codepaths.

philn replied:
Why is `gst_video_info_from_caps()` not called for `video/x-raw` caps? This looks incorrect...

philn replied:
Rephrasing... with this change, `gst_video_info_from_caps()` can actually be called for non raw caps, which triggers errors.

jameshilliard replied:
> Rephrasing... with this change, `gst_video_info_from_caps()` can actually be called for non raw caps, which triggers errors.

I didn't change that behavior, it was always called from the else block before this change:
https://github.com/WebKit/WebKit/blob/d2b5507f6b8e81ec0d5bd4de2d3d583e3fd3d199/Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp#L129-L133

jameshilliard committed:
[GStreamer][VideoCapture] fix getVideoSizeAndFormatFromCaps/getVideoResolutionFromCaps for video/x-raw mime-type caps without format
https://bugs.webkit.org/show_bug.cgi?id=242290

Reviewed by Philippe Normand.

Fixes:
0:00:02.001539663  1169 0x5577d2b74810 DEBUG             video-info video-info.c:406:gst_video_info_from_caps: parsing caps video/x-raw
0:00:02.001842884  1169 0x5577d2b74810 ERROR             video-info video-info.c:546:gst_video_info_from_caps: no format given
0:00:02.002197896  1169 0x5577d2b74810 INFO     webkitvideocapturer GStreamerVideoCapturer.cpp:136:setSize: Setting size to 640x480

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(WebCore::getVideoSizeAndFormatFromCaps):
(WebCore::getVideoResolutionFromCaps):

Canonical link: https://commits.webkit.org/252112@main
